File: tools/statusline.mjs
#!/usr/bin/env node
import fs from "node:fs";
import path from "node:path";
import os from "node:os";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const DASHBOARD_DIR = process.env.CLAUDE_DASHBOARD_DIR || path.resolve(__dirname, "..");
const LOG_PATH = path.join(DASHBOARD_DIR, ".dashboard-token-log.json");
const SIDECAR_PATH = path.join(DASHBOARD_DIR, "dashboard-tokens.js");
const PROJECTS_DIR = path.join(os.homedir(), ".claude", "projects");

function localDate(d) {
  return d.getFullYear() + "-" +
    String(d.getMonth() + 1).padStart(2, "0") + "-" +
    String(d.getDate()).padStart(2, "0");
}

function localTime(d) {
  return String(d.getHours()).padStart(2, "0") + ":" +
    String(d.getMinutes()).padStart(2, "0");
}

// JSONL 한 파일에서 토큰 합산
function sumTokens(filePath) {
  let text;
  try { text = fs.readFileSync(filePath, "utf8"); } catch { return { tokens: 0, firstTs: null }; }
  let total = 0, firstTs = null;
  for (const line of text.split("\n")) {
    const s = line.trim();
    if (!s) continue;
    let obj;
    try { obj = JSON.parse(s); } catch { continue; }
    // 첫 번째 타임스탬프 수집
    if (!firstTs) {
      const ts = obj.timestamp || (obj.message && obj.message.created_at);
      if (ts) firstTs = ts;
    }
    const usage = (obj.message && obj.message.usage) || obj.usage;
    if (usage && typeof usage === "object") {
      total += (usage.input_tokens || 0)
        + (usage.output_tokens || 0)
        + (usage.cache_creation_input_tokens || 0)
        + (usage.cache_read_input_tokens || 0);
    }
  }
  return { tokens: total, firstTs };
}

// ~/.claude/projects/ 하위 모든 JSONL 열거
function allJsonlFiles() {
  const files = [];
  if (!fs.existsSync(PROJECTS_DIR)) return files;
  for (const proj of fs.readdirSync(PROJECTS_DIR)) {
    const projDir = path.join(PROJECTS_DIR, proj);
    let entries;
    try { entries = fs.readdirSync(projDir); } catch { continue; }
    for (const f of entries) {
      if (f.endsWith(".jsonl")) {
        files.push({ sessionId: f.replace(".jsonl", ""), filePath: path.join(projDir, f) });
      }
    }
  }
  return files;
}

function loadLog() {
  try { return JSON.parse(fs.readFileSync(LOG_PATH, "utf8")); }
  catch { return { sessions: {} }; }
}

function saveLog(log) {
  try { fs.writeFileSync(LOG_PATH, JSON.stringify(log, null, 2)); } catch {}
}

function buildDaily(log) {
  const byDate = {};
  for (const s of Object.values(log.sessions)) {
    if (!s || !s.date) continue;
    byDate[s.date] = (byDate[s.date] || 0) + (s.tokens || 0);
  }
  return Object.entries(byDate)
    .map(([date, tokens]) => ({ date, tokens }))
    .sort((a, b) => (a.date < b.date ? -1 : 1));
}

function buildSessions(log) {
  return Object.entries(log.sessions)
    .map(([sid, s]) => ({
      sessionId: sid,
      date: s.date || "",
      time: s.time || "",
      tokens: s.tokens || 0,
      updatedAt: s.updatedAt || ""
    }))
    .sort((a, b) => (a.updatedAt < b.updatedAt ? 1 : -1))
    .slice(0, 100);
}

function writeSidecar(daily, sessions) {
  const payload = { generatedAt: new Date().toISOString(), daily, sessions };
  const js = "window.DASHBOARD_TOKENS = " + JSON.stringify(payload, null, 2) + ";\n";
  try { fs.writeFileSync(SIDECAR_PATH, js); } catch {}
}

function main() {
  let data = {};
  try { data = JSON.parse(fs.readFileSync(0, "utf8") || "{}"); } catch {}

  const now = new Date();
  const activeSessionId = data.session_id || null;
  const model = (data.model && data.model.display_name) || "";

  const log = loadLog();
  const allFiles = allJsonlFiles();

  for (const { sessionId, filePath } of allFiles) {
    const isActive = sessionId === activeSessionId;
    const cached = log.sessions[sessionId];
    // 완료된 세션은 캐시 재사용 (단, 토큰이 0이면 재계산)
    if (cached && cached.tokens > 0 && !isActive) continue;

    try {
      const stat = fs.statSync(filePath);
      const { tokens, firstTs } = sumTokens(filePath);
      const sessionDate = firstTs
        ? localDate(new Date(firstTs))
        : localDate(new Date(stat.birthtimeMs || stat.mtimeMs));
      const sessionTime = firstTs
        ? localTime(new Date(firstTs))
        : localTime(new Date(stat.birthtimeMs || stat.mtimeMs));

      log.sessions[sessionId] = {
        date: (cached && cached.date) || sessionDate,
        time: (cached && cached.time) || sessionTime,
        tokens,
        updatedAt: now.toISOString()
      };
    } catch {
      continue;
    }
  }

  saveLog(log);
  const daily = buildDaily(log);
  const sessions = buildSessions(log);
  writeSidecar(daily, sessions);

  // 현재 세션 토큰 상태줄 출력
  const currentSession = activeSessionId && log.sessions[activeSessionId];
  const tk = currentSession
    ? (currentSession.tokens >= 1000
        ? (currentSession.tokens / 1000).toFixed(1) + "k"
        : String(currentSession.tokens))
    : "0";
  process.stdout.write(`[${model}] 🪙 ${tk} tok`);
}

main();

File: tools/statusline.py
#!/usr/bin/env python3
"""Claude Code statusline: 토큰 합산 후 Google Sheets에 upsert하고 stdout 출력."""
import json, os, sys, urllib.request, urllib.parse, datetime

DASHBOARD_DIR = os.environ.get("CLAUDE_DASHBOARD_DIR") or os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
LOG_PATH = os.path.join(DASHBOARD_DIR, ".dashboard-token-log.json")
PROJECTS_DIR = os.path.join(os.path.expanduser("~"), ".claude", "projects")
WRITE_ENDPOINT = os.environ.get("DASHBOARD_WRITE_ENDPOINT") or \
    "https://script.google.com/macros/s/AKfycbyTF2rx3EwXUeuF_zExpbasSwmAHD0vbmsEhqyydQI6mug24B7_IfT8-ncW4xTe9gFGHQ/exec"

def local_date(dt): return dt.strftime("%Y-%m-%d")
def local_time(dt): return dt.strftime("%H:%M")

def sum_tokens(file_path):
    total = 0
    first_ts = None
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            for line in f:
                s = line.strip()
                if not s: continue
                try: obj = json.loads(s)
                except: continue
                if first_ts is None:
                    ts = obj.get("timestamp") or (obj.get("message") or {}).get("created_at")
                    if ts: first_ts = ts
                usage = (obj.get("message") or {}).get("usage") or obj.get("usage") or {}
                if isinstance(usage, dict):
                    total += (usage.get("input_tokens") or 0) + \
                             (usage.get("output_tokens") or 0) + \
                             (usage.get("cache_creation_input_tokens") or 0) + \
                             (usage.get("cache_read_input_tokens") or 0)
    except: pass
    return total, first_ts

def all_jsonl_files():
    files = []
    if not os.path.isdir(PROJECTS_DIR): return files
    for proj in os.listdir(PROJECTS_DIR):
        proj_dir = os.path.join(PROJECTS_DIR, proj)
        if not os.path.isdir(proj_dir): continue
        try: entries = os.listdir(proj_dir)
        except: continue
        for f in entries:
            if f.endswith(".jsonl"):
                files.append((f[:-6], os.path.join(proj_dir, f)))
    return files

def load_log():
    try:
        with open(LOG_PATH) as f: return json.load(f)
    except: return {"sessions": {}}

def save_log(log):
    try:
        with open(LOG_PATH, "w") as f: json.dump(log, f, indent=2)
    except: pass

def upsert_to_sheets(session_id, date, time, tokens):
    if not WRITE_ENDPOINT: return
    payload = json.dumps({"sessionId": session_id, "date": date, "time": time, "tokens": tokens})
    url = WRITE_ENDPOINT + "?action=upsertTokenSession&payload=" + \
          urllib.parse.quote(payload) + "&callback=_dummy"
    try:
        req = urllib.request.Request(url, headers={"User-Agent": "statusline/1.0"})
        urllib.request.urlopen(req, timeout=5)
    except: pass

def main():
    data = {}
    try:
        raw = sys.stdin.read()
        if raw.strip(): data = json.loads(raw)
    except: pass

    now = datetime.datetime.now()
    active_session_id = data.get("session_id")
    model = (data.get("model") or {}).get("display_name") or ""

    log = load_log()
    all_files = all_jsonl_files()

    for session_id, file_path in all_files:
        is_active = session_id == active_session_id
        cached = log["sessions"].get(session_id)
        if cached and cached.get("tokens", 0) > 0 and not is_active:
            continue
        try:
            stat = os.stat(file_path)
            tokens, first_ts = sum_tokens(file_path)
            if first_ts:
                try: dt = datetime.datetime.fromisoformat(first_ts.replace("Z", "+00:00")).astimezone()
                except: dt = datetime.datetime.fromtimestamp(stat.st_mtime)
            else:
                dt = datetime.datetime.fromtimestamp(stat.st_mtime)
            session_date = local_date(dt)
            session_time = local_time(dt)
            log["sessions"][session_id] = {
                "date": (cached or {}).get("date") or session_date,
                "time": (cached or {}).get("time") or session_time,
                "tokens": tokens,
                "updatedAt": now.isoformat()
            }
        except: continue

    save_log(log)

    # 활성 세션 upsert (fire-and-forget via thread)
    if active_session_id and active_session_id in log["sessions"]:
        s = log["sessions"][active_session_id]
        import threading
        threading.Thread(
            target=upsert_to_sheets,
            args=(active_session_id, s["date"], s["time"], s["tokens"]),
            daemon=True
        ).start()
        import time; time.sleep(0.1)  # 최소 대기로 fire

    current = log["sessions"].get(active_session_id) if active_session_id else None
    if current:
        tk = current["tokens"]
        tk_str = f"{tk/1000:.1f}k" if tk >= 1000 else str(tk)
    else:
        tk_str = "0"
    sys.stdout.write(f"[{model}] 🪙 {tk_str} tok")

if __name__ == "__main__":
    main()
